lib/Runtime/Library/JavascriptArray.inl

```diff
@@ -955,45 +955,45 @@ SECOND_PASS:
             return true;
         }
 
-        SparseArraySegment<T> *toSegment = PrepareSegmentForMemOp<T>(toStartIndex, length);
-        if (toSegment == nullptr)
+        const auto isSegmentValid = [length](Js::SparseArraySegment<T>* segment, uint32 startIndex) {
+            uint32 end, segmentEnd;
+            // Check the segment is long enough
+            return (
+                segment &&
+                !UInt32Math::Add(startIndex, length, &end) &&
+                !UInt32Math::Add(segment->left, segment->length, &segmentEnd) &&
```

Comment:
Actually, if the segment was big enough to begin with, but its length was smaller that its allocated size, your current code will call it as invalid if we try to write beyond segment->length but within segment->size.